### PR TITLE
windows: Windows client dependency code

### PR DESCRIPTION
**Pull request by dmga44**
This PR contains the necessary code to be added/modified in the common-platform wise side of the code to enable Windows client:
- build
- testing
- fix/add features

### CODE REVIEW

File: config.h.in [hunk 1]
```diff
@@ -9,6 +9,10 @@
 #define SAUNAFS_PACKAGE_VERSION_MICRO @PACKAGE_VERSION_MICRO@
 #define SAUNAFS_PACKAGE_VERSION "@PACKAGE_VERSION@"
 #define SAUNAFS_VERSHEX ((@PACKAGE_VERSION_MAJOR@)*0x10000+(@PACKAGE_VERSION_MINOR@)*0x100+(@PACKAGE_VERSION_MICRO@))
+#ifdef _WIN32
+#define WINDOWS_ClIENT_VERSION "@WINCLIENT_PACKAGE_VERSION@"
```
Comment (lgsilva3087) [inline config.h.in:13]: Is the version of the Windows Client different than the other parts of the system?

File: src/common/stat32.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * This file redefines the stat struct coming with the system for the Windows
+ * builds, for the Linux builds it just takes the coming definition with
+ * sys/stat. The target of this file is to align the stat structure with what
+ * is expecting the SaunaFS system to reduce warnings like narrow-conversion and
+ * to tackle the issue with uint16_t gid field coming with Windows default stat.
+ * Those 16 bits for the gid field causes overflow in our Windows client and
+ * must be changed.
+ */
```
Comment (lgsilva3087) [inline src/common/stat32.h:29]: Great comment. Thanks.

File: src/mount/mastercomm.cc [hunk 3]
```diff
@@ -95,6 +95,11 @@ static bool disconnect;
 static time_t lastwrite;
 static int sessionlost;
 
+#ifdef _WIN32
+uint8_t *session_flags_ = NULL;
+int *mounting_uid_ = NULL, *mounting_gid_ = NULL;
```
Comment (lgsilva3087) [inline src/mount/mastercomm.cc:100]: Please use one declaration per line and nullptr.
Also consider renaming the variables to gradually match de defined code style: mountingUid and mountingGid.

File: src/mount/mastercomm.cc [hunk 4]
```diff
@@ -806,7 +814,19 @@ int fs_connect(bool verbose) {
 			}
 #else
```
Comment (lgsilva3087) [inline src/mount/mastercomm.cc:815]: I am not familiarized with this part. Consider adding commnets explaining why is this needed for Windows to mount the filesystem in META mode.

File: src/mount/mastercomm.cc [hunk 5]
```diff
@@ -806,7 +814,19 @@ int fs_connect(bool verbose) {
...
+				}
+				*mounting_uid_ = mapalluid;
+				*mounting_gid_ = mapallgid;
+			}
+			if (*mounting_uid_ == USE_LOCAL_ID && *mounting_gid_ == USE_LOCAL_ID) {
+				fprintf(stderr, " ; users mapped to local IDs");
+			} else {
+				fprintf(stderr, " ; users mapped to %" PRIu32 ":%" PRIu32,
+				        *mounting_uid_, *mounting_gid_);
+			}
```
Comment (lgsilva3087) [inline src/mount/mastercomm.cc:829]: Anyway an error will be printed?

File: src/mount/sauna_client.cc [hunk 6]
```diff
@@ -224,6 +224,59 @@ static std::atomic<bool> gDirectIo(false);
 static uint32_t lock_request_counter = 0;
 static std::mutex lock_request_mutex;
 
+#ifdef _WIN32
```
Comment (lgsilva3087) [inline src/mount/sauna_client.cc:227]: I understand the idea, and look good enough for first version, but in the future we need to avoid implementation specifics in this part of the code. Probably the idea will be to convert the client to a class, and just extend it, but this mixture is strange and can led to extensibility issues, for instance if we plan to have another windows client based in a different library.

**Review by lgsilva3087 — COMMENTED** [review]
WIP: please see my comments.

File: src/mount/oplog.cc
Comment (uristdwarf) [inline src/mount/oplog.cc:1]: Lot's of non-relevant style changes, please revert them.

(Hopefully in the future we can run clang-format on our entire codebase)

File: src/mount/special_inode.cc [hunk 7]
```diff
@@ -41,9 +41,14 @@ const Attributes InodeStats::attr =
 	  {{'f', 0x01,0xA4, 0,0,0,0, 0,0,0,0, 0,0,0,0, 0,0,0,0, 0,0,0,0, 0,0,0,1, 0,0,0,0,0,0,0,0}};
 const Inode InodeStats::inode_ = SPECIAL_INODE_STATS;
 
-// 0x0100 == 0b100000000 == 0400
+// Win: 0x0124 == 0b100100100 == 0444
+// Other OSs: 0x0100 == 0b100000000 == 0400
```
Comment (uristdwarf) [inline src/mount/special_inode.cc:45]: By Other OS's you mean Unix systems?

File: create_win_release.bat
Comment (uristdwarf) [inline create_win_release.bat:1]: This should not be included in the open-source SaunaFS repository, since the open-source version does not officially support Windows.

**Review by uristdwarf — CHANGES_REQUESTED** [review]
Aside from a few hiccups, it's nicely done.

However I have two concerns: One is legal/business related, which I will post in the company chat and not here. The other is mixing Unix and Windows code. AFAIK this is how LizardFS did it, but I'm not sure this is the right approach. I would rather have independent common/mount libraries and tests for windows, and implementing what is exactly needed for the Windows client. Yes, we will probably have to duplicate some functionality and keep two separate tests/libraries for each OS, but in my opinion it's worth it because allows SaunaFS to focus on supporting Linux (and maybe other UNIX) systems without worrying about breaking anything Windows related. And Windows can be maintained independently. I guess my main worry is maintainability going forward.

Maybe I'm wrong and this won't cause problems long-term, but I need convincing.

File: tests/test_suites/ShortSystemTests/test_custom_goal_migration_during_delay_disconnect.sh [hunk 8]
```diff
@@ -40,7 +40,11 @@ saunafs setgoal raid_goal "${info[mount0]}"/* > /dev/null
 
 # Chunks should be replicated only on 'raid' server.
 # Replication to other servers is delayed because of disconnected chunkserver 2.
-assert_eventually_prints 20 'find_chunkserver_metadata_chunks 0 | wc -l' '5 seconds'
+if is_windows_system; then
+	assert_eventually_prints 20 'find_chunkserver_metadata_chunks 0 | wc -l' '25 seconds'
+else
+	assert_eventually_prints 20 'find_chunkserver_metadata_chunks 0 | wc -l' '5 seconds'
+fi
```
Comment (lgsilva3087) [inline tests/test_suites/ShortSystemTests/test_custom_goal_migration_during_delay_disconnect.sh:47]: Similar here, the replication speed should only be affected like this by the master and chunkservers configuration, not the client.

File: tests/test_suites/ShortSystemTests/test_file_repair_correct_only_chunk_missing.sh [hunk 9]
```diff
@@ -30,7 +30,11 @@ cd ..
 # Unmount old SaunaFS client 0:
 assert_success saunafs_mount_unmount 0
 # Mount SaunaFS client 0:
-assert_success saunafs_mount_start 0
+if is_windows_system; then
+	saunafs_mount_start 0
+else
+	assert_success saunafs_mount_start 0
+fi
```
Comment (lgsilva3087) [inline tests/test_suites/ShortSystemTests/test_file_repair_correct_only_chunk_missing.sh:37]: Please check for the differences.

File: tests/test_suites/ShortSystemTests/test_global_io_limiting_read_write.sh [hunk 10]
```diff
@@ -16,7 +19,11 @@ CHUNKSERVERS=3 \
 	MOUNT_EXTRA_CONFIG="sfscachemode=NEVER" \
 	setup_local_empty_saunafs info
 
-truncate -s1M "${info[mount0]}/read"
+if is_windows_system; then
+	FILE_SIZE=1048576 file-generate "${info[mount0]}/read"
+else
+	truncate -s1M "${info[mount0]}/read"
+fi
```
Comment (lgsilva3087) [inline tests/test_suites/ShortSystemTests/test_global_io_limiting_read_write.sh:26]: Please comment about why is different (truncate).

File: tests/tools/timeout.sh [hunk 11]
```diff
@@ -19,6 +19,11 @@ timeout_killer_thread() {
 		local value=$(($(date +%s -d "$value_string") - $(date +%s)))
 		local now_ts=$(date +%s)
 
+		# Triples test time in Windows environment
+		if is_windows_system; then
+			multiplier=$(python3 -c "print(int(5 * $multiplier))")
```
Comment (lgsilva3087) [inline tests/tools/timeout.sh:24]: Please use bash here.

**Review by lgsilva3087 — CHANGES_REQUESTED** [review]
WIP 2: Please see the suggestions, the more important are related to the differences between windows and others.

File: config.h.in (hunk 1 above)
Comment (dmga44) [reply config.h.in:13]: Yes, latest version is 1.0.4.

File: src/mount/mastercomm.cc (hunk 4 above)
Comment (dmga44) [reply src/mount/mastercomm.cc:815]: Well, I'll try to explain a bit.

The previous if was "if (!gInitParams.meta)", so I means we are not mounting in META mode, to my understanding. Anyway, the lines in between the processor directives specify different behaviors for the Windows builds and the other builds. The Windows specific code here only makes consistent the use of client side uid/gid mappings and master side uid/gid mappings (this is the master exports.cfg file). So, we could have the case where the master defines some mapping and from the client side there is also some mapping defined, and this lines set it straight:
- it overrides the client side mapping if there is some master side mapping
- tells the user about the conection parameters (specifically the uid/gid final status)

For instance, this could be a conection parameters line (bold, the uid/gid final status is bold and italic):
>$ saunafscli -D H -H 192.168.195.48
**sfsmaster accepted connection with parameters: read-write,restricted_ip ; root mapped to 0:0 ; _users mapped to local IDs_**
The service saunafscli has been started.

File: src/mount/mastercomm.cc (hunk 3 above)
Comment (dmga44) [reply src/mount/mastercomm.cc:100]: Will be updated.

Comment (dmga44) [reply src/mount/mastercomm.cc:100]: Will be updated.

File: src/mount/mastercomm.cc (hunk 5 above)
Comment (dmga44) [reply src/mount/mastercomm.cc:829]: Before these changes, it was showing users mapped to 999:999 when no master side mapping was applied. The 999 are the default values the master sends to the client when no mapping is done at all, this is confusing if take into account we are also probably doing another client side mapping.

File: src/mount/oplog.cc [hunk 12]
```diff
@@ -29,50 +29,70 @@
...
-#define OPBUFFSIZE 0x1000000
-#define LINELENG 1000
-#define MAXHISTORYSIZE 0xF00000
+#define OPBUFFSIZE          0x1000000
+#define LINELENG            1000
+#define MAXHISTORYSIZE      0xF00000
+#define TIMEDATAPREFFIXSIZE 34
 
-typedef struct _fhentry {
+    typedef struct _fhentry {
```
[the remark this answers is not in the document]
Comment (dmga44) [reply src/mount/oplog.cc:41]: Done.

File: src/mount/sauna_client.cc (hunk 6 above)
Comment (dmga44) [reply src/mount/sauna_client.cc:227]: Understood.

**Comment by dmga44** [comment]
> Aside from a few hiccups, it's nicely done.
> 
> However I have two concerns: One is legal/business related, which I will post in the company chat and not here. The other is mixing Unix and Windows code. AFAIK this is how LizardFS did it, but I'm not sure this is the right approach. I would rather have independent common/mount libraries and tests for windows, and implementing what is exactly needed for the Windows client. Yes, we will probably have to duplicate some functionality and keep two separate tests/libraries for each OS, but in my opinion it's worth it because allows SaunaFS to focus on supporting Linux (and maybe other UNIX) systems without worrying about breaking anything Windows related. And Windows can be maintained independently. I guess my main worry is maintainability going forward.
> 
> Maybe I'm wrong and this won't cause problems long-term, but I need convincing.

Well, note that there is not that much Windows specific code, and that updates in the common/mount libraries should be somehow synced with the Windows builds. That is why this PR saves us work, like bringing the new reading algorithm (if this PR was done prior that one) or the updates in the write algorithm could be easily applied everywhere. I think that in the long run, it will help.

File: tests/tools/config.sh [hunk 13]
```diff
@@ -17,7 +31,11 @@ fi
 : ${TEMP_DIR:=/tmp/SaunaFS-autotests}
 : ${LEGACY_DIR:=/tmp/SaunaFS-autotests-legacy}
 : ${SAUNAFSXX_DIR_BASE:=/tmp/SaunaFS-autotests-old}
-: ${SAUNAFS_ROOT:=$HOME/local}
+if is_windows_system; then
+	: ${SAUNAFS_ROOT:=/usr/local}
+else
+	: ${SAUNAFS_ROOT:=$HOME/local}
+fi
```
[the remark this answers is not in the document]
Comment (dmga44) [reply tests/tools/config.sh:38]: It seems to be one of my old changes, which I don't remember why I did it. Probably it didn't work the other way and just worked that way.

Now it seems to be working fine only with /usr/local. Will be updated.

[the remark this answers is not in the document]
Comment (dmga44) [reply tests/tools/config.sh:38]: It seems to be one of my old changes, which I don't remember why I did it. Probably it didn't work the other way and just worked that way.

Now it seems to be working fine only with /usr/local. Will be updated.

[the remark this answers is not in the document]
Comment (dmga44) [reply tests/tools/config.sh:38]: It seems to be one of my old changes, which I don't remember why I did it. Probably it didn't work the other way and just worked that way.

Now it seems to be working fine only with /usr/local. Will be updated.

File: tests/tools/saunafs.sh [hunk 14]
```diff
@@ -561,33 +626,58 @@ configure_mount_() {
...
+		if [ ! -d "$mount_dir" ]; then
+			sudo mkdir "$mount_dir"
+		fi
+
+		saunafs_info_[mount${mount_id}]=$mount_dir
+		saunafs_info_[mount${mount_id}_cfg]=$mount_cfg
+		saunafs_info_[mntcall${mount_id}]=$mount_call
+		saunafs_info_[mnt${mount_id}_command]="sfsmount3.exe"
+		saunafs_info_[umountcall${mount_id}]=$umount_call
+		saunafs_info_[mount_win_path${mount_id}]=${mount_big_dir_letter}:
```
[the remark this answers is not in the document]
Comment (dmga44) [reply tests/tools/saunafs.sh:655]: Added windows_prepare_mount_ and unix_prepare_mount_ functions for each behavior.

File: tests/tools/config.sh (hunk 13 above)
[the remark this answers is not in the document]
Comment (dmga44) [reply tests/tools/config.sh:38]: It seems to be one of my old changes, which I don't remember why I did it. Probably it didn't work the other way and just worked that way.

Now it seems to be working fine only with /usr/local. Will be updated.

File: tests/tools/test.sh [hunk 15]
```diff
@@ -144,18 +146,30 @@ mass_chmod_777() {
...
+		# Search for all fuse filesystems mounted by user saunafstest and umount them
+		local uid=$(id -u saunafstest)
+		while list_of_mounts=$(cat /proc/mounts | awk '$3 ~ /^fuse/' | grep "user_id=$uid[^0-9]"); do
+			echo "$list_of_mounts" | awk '{print $2}' | xargs -r -d'\n' -n1 fusermount -u || sleep 1
+			if ((++retries == 30)); then
+				echo "Can't unmount: $list_of_mounts" >&2
+				break
+			fi
+		done
+	fi
```
[the remark this answers is not in the document]
Comment (dmga44) [reply tests/tools/test.sh:172]: Added windows_unmount_fs and unix_unmount_fs functions for each behavior. Please note these lines only unmount the avilable filesystems. The remaining of the test_cleanup function is common.

File: tests/test_suites/ShortSystemTests/test_auto_recovery_repair.sh [hunk 16]
```diff
@@ -13,6 +13,10 @@ mkdir dir
 saunafs setgoal 3 dir
 FILE_SIZE=$(( 4 * SAUNAFS_CHUNK_SIZE )) file-generate dir/file
 
+if is_windows_system; then
+	sleep 5
```
[the remark this answers is not in the document]
Comment (dmga44) [reply tests/test_suites/ShortSystemTests/test_auto_recovery_repair.sh:17]: These tests were doing some IO operation of over 1mb. It seems that under WSL1 the CS are running very slow and there are a few of them in each of these cases. Thus, to succeed on the following checks (like checking if there is some amount of chunks of this or that type) done in those tests we must wait for the CS. It applies to several other tests.

I chose 5 seconds cuz it was the least time it was passing everywhere, I'll add as you said a function called ```wait_chunkservers``` which just sleeps 5s. 

Also note that I'm not interested in checking the speed or well-behavior of the CSs, and I suppose them to be perfect.

File: src/mount/oplog.cc
Comment (dmga44) [reply src/mount/oplog.cc:1]: Done.

File: src/mount/special_inode.cc (hunk 7 above)
Comment (dmga44) [reply src/mount/special_inode.cc:45]: Well, basicly yes. I was referring to other than Windows OSs.

File: create_win_release.bat
Comment (dmga44) [reply create_win_release.bat:1]: I agree it would look better somewhere else. Anyway I would probably finish with that script there in my local clone. This makes me think I should add the soft/hardlinks created to the gitignore file.

File: tests/test_suites/ShortSystemTests/test_auto_recovery_repair.sh (hunk 16 above)
[the remark this answers is not in the document]
Comment (dmga44) [reply tests/test_suites/ShortSystemTests/test_auto_recovery_repair.sh:17]: Please check this comment: https://github.com/leil-io/saunafs/pull/24#discussion_r1522590818 .

Let me add that those sleeps were added when improving WindowsClientTests "stability". Sometimes those cases were failing and the final solutions were the ones developed.

File: tests/test_suites/ShortSystemTests/test_auto_recovery_repair.sh [hunk 17]
```diff
@@ -37,6 +46,11 @@ assert_success rm "$chunk"
 
 # Update first and second chunk of the file, this will change it's version to 3 on CS 2.
 dd if=/dev/zero of=dir/file conv=notrunc bs=32KiB count=$((2*1024 + 10))
+
+if is_windows_system; then
+	sleep 5
```
[the remark this answers is not in the document]
Comment (dmga44) [reply tests/test_suites/ShortSystemTests/test_auto_recovery_repair.sh:51]: Well, you may consider this: the whole WindowsClientTests suite takes around 100mins to run in my PC. All the "sleep 5" commands takes in total 75s. Tests executions under WSL1 are already very slow, both tests were performed in my laptop:
- ```[       OK ] SanityChecks.test_write_and_read (70694 ms)``` ---> WSL1
- ```[       OK ] SanityChecks.test_write_and_read (13171 ms)``` ---> WSL2
That is why I thought it did not make much change in time if it waits 5s more.

File: tests/test_suites/ShortSystemTests/test_cs_failure_recovery_speed.sh [hunk 18]
```diff
@@ -9,15 +9,25 @@ CHUNKSERVERS=4 \
...
+fi
 
 # Stop one of chunkservers which has one copy of each chunk and
 # immediately start overwriting all the chunks.
-kill -s SIGSTOP "$(saunafs_chunkserver_daemon 0 test 2>&1 | sed 's/.*pid: //')"
+if is_windows_system; then
+	saunafs_chunkserver_daemon 0 stop
+else
+	kill -s SIGSTOP "$(saunafs_chunkserver_daemon 0 test 2>&1 | sed 's/.*pid: //')"
+fi
```
[the remark this answers is not in the document]
Comment (dmga44) [reply tests/test_suites/ShortSystemTests/test_cs_failure_recovery_speed.sh:22]: Well, there are changes in fact. The main idea behind this change is that under WSL1 looks like some signals are successfully processed or something like that because the saunafs masters and chunkservers "cannot find themselves". 

For instance, using two times the command "sfschunkserver -d" creates two processes rather than stopping the first one. The second chunkserver process ends up indefinately asking to connect to master.

The current status of the PR tries to tackle this issue with the function ```windows_server_aux``` from the saunafs.sh file. The idea is to directly find the process and send the signal with the ```kill``` command. It only implemented the reload, kill, stop and start commands, though start might not be strightforward. The case is that I could not get the test command running.

This test case, in its oriignal version uses ```saunafs_chunkserver_daemon 0 test``` which is not working right now, and just left the command in the much simpler stop command.

File: tests/test_suites/ShortSystemTests/test_custom_goal_migration_during_delay_disconnect.sh (hunk 8 above)
Comment (dmga44) [reply tests/test_suites/ShortSystemTests/test_custom_goal_migration_during_delay_disconnect.sh:47]: This is the same issue with speed as mentioned here: https://github.com/leil-io/saunafs/pull/24#discussion_r1522690499.

File: tests/test_suites/ShortSystemTests/test_file_repair_correct_only_chunk_missing.sh (hunk 9 above)
Comment (dmga44) [reply tests/test_suites/ShortSystemTests/test_file_repair_correct_only_chunk_missing.sh:37]: Well, I don't remember right now but the result of leaving the ```assert_success``` is the testing process hanged and then TL. I'll probably look further  into this issue later.

File: tests/test_suites/ShortSystemTests/test_global_io_limiting_read_write.sh (hunk 10 above)
Comment (dmga44) [reply tests/test_suites/ShortSystemTests/test_global_io_limiting_read_write.sh:26]: Looks like old client version did not support it. Probably at some point I thought it was WSL1 fault of not supporting truncate (like chmod and chown) in a drive mounted using drvfs.

Current client version passed the test with the truncate command. Will revert it.

File: tests/tools/timeout.sh (hunk 11 above)
Comment (dmga44) [reply tests/tools/timeout.sh:24]: Have to say that other calculation in the same script also used python, and I based on that one.

I wil update the comment, and the two calculations.

File: CMakeLists.txt [hunk 19]
```diff
@@ -335,6 +335,17 @@ if(ZLIB_FOUND)
...
 include_directories(src) # Allow paths relative to src, eg. #include common/crc.h
 link_directories(${LINK_DIRECTORIES} ${FUSE_LIBRARY_DIR})
+
+if(MINGW)
+  option(ENABLE_WINFSP_WINCLIENT  "Enable winfsp library based windows client build"    ON)
+
+  if(ENABLE_WINFSP_WINCLIENT)
+    set(WINFSP_FUSE3_INCLUDE_DIR  "C:/Program Files (x86)/WinFsp/inc/fuse3"             CACHE STRING "winfsp installation /inc directory")
+    set(WINFSP_FUSE_DLL_FOR_ARCH  "C:/Program Files (x86)/WinFsp/bin/winfsp-x64.dll"    CACHE STRING "winfsp-fuse dll for specific arch")
+    set(WINFSP_FUSE_LIBRARY  "C:/Program Files (x86)/WinFsp/lib/winfsp-x64.lib"    CACHE STRING "winfsp-fuse lib for specific arch")
```
[the remark this answers is not in the document]
Comment (lgsilva3087) [reply CMakeLists.txt]: I see it was done.

File: config.h.in (hunk 1 above)
Comment (lgsilva3087) [reply config.h.in:13]: We need to agree on how to handle this, because it can be confusing for the end user. Please, see the picture, the package version might be 1.*, but it installs client 4.*.
image

File: config.h.in [hunk 20]
```diff
@@ -9,6 +9,10 @@
...
+#define WINFSP_LIBRARY_VERSION "@WINFSP_VERSION@"
```
Comment (lgsilva3087) [inline config.h.in:14]: Looks like these macros are not used in this project. Please consider moving them to the windows submodule.

File: src/mount/mastercomm.cc (hunk 3 above)
Comment (lgsilva3087) [reply src/mount/mastercomm.cc]: I see it was already done.

File: src/mount/oplog.cc [hunk 21]
```diff
@@ -76,23 +75,23 @@ static inline void oplog_put(uint8_t *buff, uint32_t leng) {
 	}
 #endif
 	uint32_t bpos;
-	if (leng > OPBUFFSIZE) { // just in case
-		buff += leng - OPBUFFSIZE;
-		leng = OPBUFFSIZE;
+	if (leng>OPBUFFSIZE) {  // just in case
```
Comment (rolysr) [inline src/mount/oplog.cc:78]: Do you have .clang-format file correctly configured in your environment in order to keep proper spacing around operators, function arguments etc. for code readability? Consider to check the main changes made on oplog.cc file to check this code format issue.

File: tests/tools/test.sh [hunk 22]
```diff
@@ -142,33 +142,41 @@ mass_chmod_777() {
...
+	local uid=$(id -u saunafstest)
+	while list_of_mounts=$(cat /proc/mounts | awk '$3 ~ /^fuse/' | grep "user_id=$uid[^0-9]"); do
+		echo "$list_of_mounts" | awk '{print $2}' | xargs -r -d'\n' -n1 fusermount -u || sleep 1
+		if ((++retries == 30)); then
+			echo "Can't unmount: $list_of_mounts" >&2
+			break
+		fi
+	done
+}
+
```
Comment (rolysr) [inline tests/tools/test.sh]: Consider to add a little doc comment to the new added functions to the .sh files on tests/tools folder. Maybe it would be better if we keep all these util procedures with a minimum doc as there is present is some other ones.

File: tests/tools/time.sh [hunk 23]
```diff
@@ -22,6 +22,10 @@ wait_for() {
 	return 1
 }
 
+wait_chunkservers() {
+	sleep 5
```
Comment (rolysr) [inline tests/tools/time.sh:26]: In addition to adding a small comment about what this function is used for, it would be good to specify why 5 was selected as the chunkservers' wait value.

File: tests/test_suites/ShortSystemTests/test_custom_goals_when_creating_new_chunks.sh [hunk 24]
```diff
@@ -1,4 +1,8 @@
-timeout_set "1 minute"
+if is_windows_system; then
+	timeout_set "2 minute"
```
Comment (rolysr) [inline tests/test_suites/ShortSystemTests/test_custom_goals_when_creating_new_chunks.sh:2]: Could you consider to change timeout value to "2 minutes"?

File: src/mount/oplog.cc (hunk 21 above)
Comment (uristdwarf) [reply src/mount/oplog.cc:78]: If the line changes are not relevant for what the commit is attempting to achieve, they should not be changed because this pollutes the blame history with non-relevant changes that will be difficult differentiate. This reverses a previous commit changes (as I requested for the previous reason) https://github.com/leil-io/saunafs/pull/24#discussion_r1521748912 

Hopefully before the next release we could do the style changes (even if they are disruptive) and filter them out in [.git-blame-ignore-revs](https://github.com/leil-io/saunafs/blob/main/.git-blame-ignore-revs).

File: tests/tools/time.sh [hunk 25]
```diff
@@ -22,6 +22,10 @@ wait_for() {
 	return 1
 }
 
+wait_chunkservers() {
```
Comment (uristdwarf) [inline tests/tools/time.sh:25]: The condition should not be repeated everywhere again, rather it should be in this function (unless it's not possible?)

```suggestion
wait_if_windows() { # Or windows_wait_chunkservers
    if is_windows_system; then
        sleep 5
    fi
}
```
(Please don't commit directly since it uses spaces)

Also as @rolysr mentions, documentation on why this is needed and why 5 seconds exactly.

**Review by uristdwarf — APPROVED** [review]
Approved, but note two things:

1. We need to think about making the client library more extensible, to avoid contaminating the core code with functionality for other platforms (maybe like how chunkserver was done) to avoid this in the future. Hopefully no more Windows code are needed for a while, and we can start working on making this extensible.

2. Tests are also iffy. I know understand the reason why it's needed, but the way it is done is unstable (relying on WSL) and makes the tests designed for Linux systems dependent on Windows systems as well. We need to think some better ways to do this, so maybe these will be removed in the future. I will keep for now though because it's really needed.

File: src/mount/special_inode.cc (hunk 7 above)
Comment (uristdwarf) [reply src/mount/special_inode.cc:45]: I guess this is more of a nit than anything (so I don't care), but there are more operating systems than Unix systems like Linux or macOS.

https://eylenburg.github.io/pics/Eylenburg_Operating_System_Timeline_Family_Tree.svg

File: tests/test_suites/ShortSystemTests/test_auto_recovery_repair.sh (hunk 16 above)
[the remark this answers is not in the document]
Comment (uristdwarf) [reply tests/test_suites/ShortSystemTests/test_auto_recovery_repair.sh]: Again, the testing feels kinda faulty because it's not really testing the actual environment where SaunaFS is being used.

But please add the reasoning as comments there, for future developers.

**Comment by dmga44** [comment]
> makes the tests designed for Linux systems dependent on Windows systems as well

@uristdwarf Why? I don't understand this. On all the other I agree with you, but all the changes introduced in the core and in the tests are only used inside is_windows_system ifs. So, execution of test under Linux has nothing to do with Windows platform, it just checks if its WSL and if that´s not the case it continues normally. So, what do you mean here?

**Comment by uristdwarf** [comment]
> @uristdwarf Why? I don't understand this. On all the other I agree with you, but all the changes introduced in the core and in the tests are only used inside is_windows_system ifs. So, execution of test under Linux has nothing to do with Windows platform, it just checks if its WSL and if that´s not the case it continues normally. So, what do you mean here?

I guess the better term would be "reliant on WSL".

Imagine if we added some new tests that doesn't work on WSL, however it works perfectly on Linux and thus passes. Then we later find out it doesn't work on WSL and someone needs to go and fix it by adding another workaround. Or we need to use some command line tool that for whatever reason doesn't work on WSL, again another if_windows workaround that needs to be added. It's the same thing with the code, potential for breakage is large, and adding hacks to fix it isn't what I would consider a good solution.

I want the tests (and SaunaFS) to work on Ubuntu, I don't care if it doesn't work on Arch or Fedora or Debian. WSL for me is just another distribution and while I understand the reasoning why we need it (to test the Windows client on our current tests), it still gives us another distribution to support and maintain our tests for (and potentially in the future, we may have to support something like macOS or Fedora as well, so even more hacks).

So not really dependent on Windows, but dependent on a distribution (that is heavily integrated with Windows).

**Comment by uristdwarf** [comment]
Maybe I'm overthinking this (and again I'm not against including this code for now), but I feel we need a better solution in the end for testing various platforms, and currently I don't really think extending the tests for the various platforms is a good solution because we need to maintain the tests for every platform we add.
